src/aspire/source/relion.py

```diff
@@ -1,84 +1,138 @@
+        '_offset_y':    '_rlnOriginY',
+        '_state':       '_rlnClassNumber',
+        '_angle_0':     '_rlnAngleRot',
+        '_angle_1':     '_rlnAngleTilt',
+        '_angle_2':     '_rlnAnglePsi',
+        '_amplitude':   '_amplitude',
+        '_voltage':     '_rlnVoltage',
+        '_defocus_u':   '_rlnDefocusU',
+        '_defocus_v':   '_rlnDefocusV',
+        '_defocus_ang': '_rlnDefocusAngle',
+        '_Cs':          '_rlnSphericalAberration',
+        '_alpha':       '_rlnAmplitudeContrast'
+    }
 
-        return df
+    _metadata_types = {
+        '_rlnVoltage': float,
+        '_rlnDefocusU': float,
+        '_rlnDefocusV': float,
+        '_rlnDefocusAngle': float,
+        '_rlnSphericalAberration': float,
+        '_rlnDetectorPixelSize': float,
+        '_rlnCtfFigureOfMerit': float,
+        '_rlnMagnification': float,
+        '_rlnAmplitudeContrast': float,
+        '_rlnImageName': str,
+        '_rlnOriginalName': str,
+        '_rlnCtfImage': str,
+        '_rlnCoordinateX': float,
+        '_rlnCoordinateY': float,
+        '_rlnCoordinateZ': float,
+        '_rlnNormCorrection': float,
+        '_rlnMicrographName': str,
+        '_rlnGroupName': str,
+        '_rlnGroupNumber': str,
+        '_rlnOriginX': float,
+        '_rlnOriginY': float,
+        '_rlnAngleRot': float,
+        '_rlnAngleTilt': float,
+        '_rlnAnglePsi': float,
+        '_rlnClassNumber': int,
+        '_rlnLogLikeliContribution': float,
+        '_rlnRandomSubset': int,
+        '_rlnParticleName': str,
+        '_rlnOriginalParticleName': str,
+        '_rlnNrOfSignificantSamples': float,
+        '_rlnNrOfFrames': int,
+        '_rlnMaxValueProbDistribution': float
+    }
 
-    def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_files=False, max_rows=None):
+    @classmethod
+    def starfile2df(cls, filepath, block_index_or_name=0, loop_index=0, ignore_missing_files=False, max_rows=None):
+
+        dirpath = os.path.dirname(filepath)
+        df = Starfile(filepath)[block_index_or_name][loop_index].data
+        column_types = {name: cls._metadata_types.get(name, str) for name in df.columns}
+        df = df.astype(column_types)
+
+        # Rename fields to standard notation
+        reverse_metadata_aliases = {v: k for k, v in cls._metadata_aliases.items()}
+
+        df = df.rename(reverse_metadata_aliases, axis=1)
+
+        # TODO: Check behavior if this is a single mrc file (no '@')
+        _index, df['__mrc_filename'] = df['_image_name'].str.split('@', 1).str
+        df['__mrc_index'] = pd.to_numeric(_index)
+
+        # Adding a full-filepath field to the Dataframe helps us save time later
+        # Note that os.path.join works as expected when the second argument is an absolute path itself
+        df['__mrc_filepath'] = df['__mrc_filename'].apply(lambda filename: os.path.join(dirpath, filename))
+
+        # Helper function to check if a file exists
+        def file_exists(filepath):
```

Comment:
I've routinely used this in development so far, where I don't want to fetch all mrcs files, but just a small subset of them (and don't want to edit the STAR file either, since there is no inherent guarantee on the ordering of rows), but just direct things to do a best-effort approach.

The default value of `ignore_missing_files` is `False` of course.